pkg/operator/configobserver/cloudconfig/sync_cloud_config.go

```diff
@@ -0,0 +1,121 @@
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+)
+
+const (
+	configNamespace                 = "openshift-config"
+	machineSpecifiedConfigNamespace = "openshift-config-managed"
+	machineSpecifiedConfig          = "kube-cloud-config"
+)
+
+var validCloudProviders = sets.NewString("aws", "azure", "gce", "openstack", "vsphere")
+
+// InfrastructureLister lists infrastrucre information and allows resources to be synced
+type InfrastructureLister interface {
+	InfrastructureLister() configlistersv1.InfrastructureLister
+	ResourceSyncer() resourcesynccontroller.ResourceSyncer
+	ConfigMapLister() corelisterv1.ConfigMapLister
+}
+
+// NewCloudConfigSyncer returns a new cloudprovider dependent observer for syncing cloud config per provider
+// for controller-manager and api-server.
+func NewCloudConfigSyncer(targetNamespaceName string) configobserver.ObserveConfigFunc {
+	cloudObserver := &cloudConfigSyncer{
+		targetNamespaceName: targetNamespaceName,
+	}
+	return cloudObserver.SyncCloudConfig
+}
+
+type cloudConfigSyncer struct {
+	targetNamespaceName string
+}
+
+// ObserveCloudProviderNames observes the cloud provider from the global cluster infrastructure resource.
+func (c *cloudConfigSyncer) SyncCloudConfig(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(InfrastructureLister)
+
+	infrastructure, err := listers.InfrastructureLister().Get("cluster")
+	if err != nil {
+		if errors.IsNotFound(err) {
+			recorder.Warningf("ObserveCloudProviderNames", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		}
+		return existingConfig, []error{err}
+	}
+
+	cloudProvider := getPlatformName(infrastructure.Status.Platform, recorder)
+
+	// we copy cloudprovider configmap values only for some cloud providers.
+	if !validCloudProviders.Has(cloudProvider) {
+		return existingConfig, nil
+	}
+
+	sourceCloudConfigMap := infrastructure.Spec.CloudConfig.Name
+	sourceCloudConfigNamespace := configNamespace
+
+	// If a managed cloud-provider config is available, it should be used instead of the default. If the configmap is not
+	// found, the default values should be used.
+	if _, err = listers.ConfigMapLister().ConfigMaps(machineSpecifiedConfigNamespace).Get(machineSpecifiedConfig); err == nil {
+		sourceCloudConfigMap = machineSpecifiedConfig
+		sourceCloudConfigNamespace = machineSpecifiedConfigNamespace
+	} else if !errors.IsNotFound(err) {
+		return existingConfig, []error{err}
+	}
+
+	sourceLocation := resourcesynccontroller.ResourceLocation{
+		Namespace: sourceCloudConfigNamespace,
+		Name:      sourceCloudConfigMap,
+		Provider:  cloudProvider,
+	}
+
+	if err := listers.ResourceSyncer().SyncConfigMap(
```

Comment:
an observer never does any actions like this. It must be free of side-effect of any kind.

Reply:
iiuc, it seems like this behavior is existent in the current implementation, https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L105

are you saying we should refactor to remove this behavior?

Reply:
you need a controller or add the configmap to the existing resource syncers

Reply:
@stts It is already there, as the observer will be called [here](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L84-L98) and [here](https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L119-L152) the informers list contain a [list](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L37-L38) of namespaces where the configMap is [watched](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L55)